### PR TITLE
65 mock middleware

### LINKED ISSUE #65: implement Mock Middleware
Implement a middleware, which can match an incoming options-parameter to a list of handlers. First handler to respond with a truthy value, will return a factorized value. If no matching handler is found, continue normally in the stack by calling next(...)

### PR DESCRIPTION
**Pull request by cyberhck**
fixes #65 
- Add Mock middleware
- Add tests for Mock middleware
- Question (do we start doing documentation too?)
- Need help: do we declare our own Response? Is it same as IResponse?

### CODE REVIEW

File: src/middlewares/Mock.spec.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    });
+
+    describe("addHandler", () => {
+        it("is a a function", () => {
+            expect(typeof (new Mock()).addHandler).toBe("function");
+        });
+
+        it("is able to add handler to array", () => {
+            const obj = new Mock();
+            obj.addHandler({match: undefined as any, delay: 5, factory: undefined as any});
```
Comment (cyberhck) [inline src/middlewares/Mock.spec.ts]: Not sure how to test if it actually added the handler

**Comment by cyberhck** [comment]
@wmathes any changes you'd like here?

File: src/middlewares/Fetch.ts [hunk 2]
```diff
@@ -1,7 +1,21 @@
-import {IRequest, IResponse} from "../Client";
+import {IRequest, IResponse, IResponseHeaders} from "../Client";
 import {combineUrlWithBaseUrl, objectToQueryString} from "../internal/formatting";
 import {IMiddleware} from "../Stack";
 
+export type FetchResponse = {
+    new(body: any, init: any): FetchResponse;
+    body: string;
+    bodyUsed: boolean;
+    headers: IResponseHeaders
```
Comment (wmathes) [inline src/middlewares/Fetch.ts]: Uhm when refactoring this to FetchResponse. I assume there will be no more IRequest Objects be present... EVER :P 

TLDR: Remove the IResponse interface. All occurrences should be replaced by FetchResponse. Same for the IRequestHeaders.

File: src/middlewares/Mock.spec.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
+import {Client, IResponse} from "../Client";
+import {Mock} from "./Mock";
+describe("Mock", () => {
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: LF please

File: src/middlewares/Mock.spec.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                    match: undefined as any,
+                    factory: undefined as any
+                }
+            ]);
+            expect(obj).toBeDefined();
+        });
+    });
+
+    describe("addHandler()", () => {
+        it("is a a function", () => {
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: typo

File: src/middlewares/Mock.spec.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+describe("Mock", () => {
+    it("is defined", () => {
+        expect(Mock).toBeDefined();
+    });
+
+    describe("constructor()", () => {
+        it("accepts IHandler array in constructor", () => {
+            const obj = new Mock([
+                {
+                    delay: 0,
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: delay is optional, remove it. No need to clutter the test file. 

Also... is this test really checking what we want? Ideas?

File: src/middlewares/Mock.spec.ts (hunk 1 above)
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: Hmm that's a valid problem.

How about the following:
1) rename addHandler to registerHandler, check if the reference is already in the list when adding. return true on success.
2) add a unregisterHandler(handler) to remove a handler. return true on success
3) add resetHandlers() to remove all registered handlers.
 
Maybe this is a bit too much, but maybe its an interesting scenario.

Otherwise for testing the add function: 
1) add something with a useful function
2) invoke process() 
3) see if your handler actually got called.

File: src/middlewares/Mock.spec.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    describe("process()", () => {
+        it("is a function", () => {
+            expect(typeof (new Mock()).process).toBe("function");
+        });
+
+        it("returns mocked data", (done) => {
+            const obj = new Mock();
+            obj.addHandler(
+                {
+                    match: (): boolean => {
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: simplify to `=> true`

File: src/middlewares/Mock.spec.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                        return true;
+                    },
+                    delay: 5,
+                    factory: (): IResponse<any> => {
+                        return Mock.jsonResponse({user: 1, post: "example post"});
+                    }
+                }
+            );
+            const client = new Client({baseUrl: "https://jsonplaceholder.typicode.com"});
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: Why do we use Client here? Just use an Mock-instance and call process()

File: src/middlewares/Mock.spec.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                res.json().then((post) => {
+                    expect(post.post).toBe("example post");
+                    expect(post.user).toBe(1);
+                    done();
+                });
+            });
+        });
+
+        it("calls next when no mocks match", () => {
+            const client = new Client({baseUrl: "https://api.example.com"});
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: same as above: No need to use Client, use a Mock-instance instead.

File: src/middlewares/Mock.spec.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            const obj = new Mock(
+                [{
+                    match: (): boolean => {
+                        return false;
+                    },
+                    factory: (): IResponse<any> => {
+                        return Mock.jsonResponse({});
+                    }
+                }]
+            );
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: simplify :)

File: src/middlewares/Mock.spec.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    describe("jsonResponse()", () => {
+        it("returns an object as a mock", (done) => {
+            const client = new Client({baseUrl: "https://api.example.com"});
+            const obj = new Mock(
+                [{
+                    match: (): boolean => {
+                        return true;
+                    },
+                    delay: 5,
+                    factory: (): Promise<IResponse<any>> => {
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: Unnecessarily complicated. Don't use Client or promise wrapping.

Just call the method and look at it's return type. Saves 20 lines and more useful ^^

File: src/middlewares/Mock.spec.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            client.process({url: "/posts/1"}).then((res) => {
+                res.json().then((post) => {
+                    expect(post.user).toBe(1);
+                    expect(post.post).toBe("example post");
+                    done();
+                });
+            });
+        });
+
+        it("uses 204 status code if used with null", (done) => {
```
Comment (wmathes) [inline src/middlewares/Mock.spec.ts]: Same as above. No need any Client or Mock instances.

File: src/middlewares/Mock.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+        // Encode data
+        const stream: string|undefined = (data === undefined || data === null) ? undefined : JSON.stringify(data);
+
+        // Create Data
+        const init = Object.assign(
+            {},
+            {
+                status: data === undefined || data === null ? 204 : 200,
+                headers: {
```
Comment (wmathes) [inline src/middlewares/Mock.ts]: Please research: I think the Content-Type header is omitted for 204 (because No-Content implies no content type)

File: src/middlewares/Mock.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        // Return Promises directly
+        if (response instanceof Promise) {
+            return response;
+        }
+
+        // Wrap result in promise and resolve after a short delay
+        const delay = handler.delay || handler.delay === 0
+            ? handler.delay
+            : this.defaultDelay;
+        return new Promise((resolve) => {
```
Comment (wmathes) [inline src/middlewares/Mock.ts]: Maybe it's elegant to move the self-resolving promise part to an extra function. I'm thinking

`public static resolvingPromise(value: any, delay: number)`

Worthwhile?

File: src/middlewares/Mock.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
+import {IRequest, IResponse} from "../Client";
+import {IMiddleware} from "../Stack";
+import {FetchResponse} from "./Fetch";
+
+export interface IMockHandler<IN, OUT> {
+    match: (options: IN) => boolean|undefined;
+    factory: (options: IN) => OUT;
```
Comment (wmathes) [inline src/middlewares/Mock.ts]: Should we rename `factory` to something else? resultFn, resultFactory, .... ? The term factory is biased.

File: src/Service.spec.ts [hunk 15]
```diff
@@ -1,43 +1,46 @@
...
+            // Has the baseUrl been correctly passed on to the client?
+            expect(defaultOptions.baseUrl).toBe(baseUrl);
+        });
 
-    it("accepts a Client-instance", () => {
-        const baseUrl = "http://api.example.com";
-        const cl = new Client({baseUrl} as IRequest);
-        const obj = new Service(cl);
-        const defaultOptions = obj.client.defaultOptions;
+        it("accepts a Client-instance", () => {
```
Comment (cyberhck) [inline src/Service.spec.ts]: d8cbde8 solves this one

File: src/middlewares/Mock.spec.ts (hunk 1 above)
Comment (cyberhck) [inline src/middlewares/Mock.spec.ts]: I used jasmine's `spyOn` instead.

File: src/middlewares/Mock.ts (hunk 14 above)
Comment (cyberhck) [inline src/middlewares/Mock.ts]: resultFactory sounds better :)
